Refactor backend handling

johananl opened the pull request:
This PR simplifies the `Backend` interface and cleans up the implementations.

johananl committed:
Refactor backend handling

- Simplify Backend interface.
  - Rather than relying on the caller to call LoadConfig, expose new
    functions for creating a backend configuration and a backend
    struct. HCL unmarshaling happens inside NewConfig now so that
    it's impossible to construct an "unloaded" backend.
  - Replace Render() with String(). The actual rendering happens
    at backend construction time and the result is stored for
    retrieval by String().
  - Remove Validate() from the interface. Instead, validate the
    the config at creation time.
- Make region a required parameter. Ensuring that a region is
  specified only when no credentials file is specified doesn't make
  sense. A region is always required by the S3 Terraform backend. In
  addition, the AWS_SHARED_CREDENTIALS_FILE and AWS_DEFAULT_REGION
  env vars aren't documented in the S3 backend reference guide. We
  might as well just get rid of these.
- Get rid of the "registration" pattern. Storing all supported
  backend types in a global variable doesn't seem useful. Instead,
  each implementation of the Backend interface can construct a
  concrete type which implements the interface and return it.
- Fix whitespace problems in backend file.

invidian reviewed (changes requested):
Hmm, I don't fully agree with the changes done in this PR. Please see my comments with explanation.

invidian reviewed (commented):
Also, there are tests missing. I think part of the refactoring should be adding tests to make sure new code is easily testable.

johananl commented on pkg/backend/s3/template.go:
```diff
@@ -14,16 +14,16 @@
-  {{- end }}
-}
-`
+var backendConfigTmpl = `terraform {
```
Do we have these other things now? If not, it might be better to refactor when we do so that it's clear how to implement this.

invidian replied:
I'd say yes, as the experimental PR for Terraform 0.13 is opened (#824) and I'm in the process of adding a CI support for it.

Also, with such implementation, you're changing the behavior (IIUC), that now backend configuration will always be present, regardless whether it was actually configured by the user or not. Currently implementation skips writing the file if it's empty.

On the other hand, I see now that it is permitted to have multiple `terraform {}` blocks. It is only not permitted to have multiple `terraform {}` blocks with `required_providers {}` sub-block. But I'm not sure if we want to have multiple ones.

johananl replied:
>Also, with such implementation, you're changing the behavior (IIUC), that now backend configuration will always be present, regardless whether it was actually configured by the user or not. Currently implementation skips writing the file if it's empty.

Nope, we're still checking: https://github.com/kinvolk/lokomotive/blob/9824ce7977b83fc74ae726d62fa3507ec01ee00c/pkg/terraform/util.go#L36-L39

We could probably improve the way we render `cluster.tf` so that we collect data from multiple sources (platform, backend etc.) and the render `cluster.tf` "in one go", however doing so exceeds the scope of this PR.

johananl replied:
I'll undo the `terraform {}` part and look for a good way to deal with the contents of this block.

invidian replied:
>I'll undo the terraform {} part and look for a good way to deal with the contents of this block.

Thanks, I guess that make sense :+1: 

>I think that pkg/terraform/util.go is a terrible place for writing the backend file 

To be honest, that depends. creating this file is actually the Terraform-related functionality, so it loosely belongs to `terraform` package.

johananl replied:
I don't think that the `terraform` package should be writing files at all.

invidian replied:
Well, terraform package could take Terraform code as an input, file is a format used by Terraform UI then :smile:

johananl replied:
I've undone the `terraform {}` block change. We now render things they way we did until now minus some whitespace problems I've fixed.

As for a better mechanism for rendering the `terraform {}` block and possibly the entire root module - I don't want to do it in this PR as it will quickly grow into yet another big refactoring journey. I'm already touching the root module generation in the platform refactoring PR (which I'm still struggling to get to) so let's see how things look after that.

johananl commented on pkg/backend/s3/s3.go:
```diff
@@ -15,61 +15,57 @@
 
-// Validate validates the s3 backend configuration.
-func (s *s3) Validate() error {
-	if s.Bucket == "" {
```
In this case it seems that "validate" means "ensure a bucket and a key were provided". If that's all we do, why not enforce this via HCL?

As for HCL vs. other formats - that's a valid concern but a purely theoretical in this context. I'm fine with having a more complex validation but IMO we should only have it where it's necessary (e.g. when we need to check more than "was X specified?").

invidian replied:
> If that's all we do, why not enforce this via HCL?

Because IMO it doesn't enforce a thing, see #239 for more details.

>As for HCL vs. other formats - that's a valid concern but a purely theoretical in this context. I'm fine with having a more complex validation but IMO we should only have it where it's necessary (e.g. when we need to check more than "was X specified?").

This is important and not theoretical, as the test for required variables shouldn't require getting HCL parser, context etc etc. Just a struct and a validation function should be enough to test validation.

johananl replied:
It is theoretical because we don't support other config formats at the moment :shrug:

Anyway, I'll re-add the validation.

johananl replied:
Done.

invidian replied:
Thanks :)

johananl commented on pkg/backend/local/template.go:
```diff
@@ -14,10 +14,10 @@
 
 package local
 
-var backendConfigTmpl = `
```
Yes, I was thinking about this, too. Didn't do it in the end to try and keep things simple. Frankly I think the whole backend abstraction might be an overkill and seems to have been copy-pasted from the platform abstraction. I'll take another look at the whole thing and see what I can improve.

johananl replied:
Actually in the current state including the `backend {}` block in `cluster.tf` would mean duplicating the block in every platform :confused:

johananl commented on cli/cmd/cluster.go:
```diff
@@ -62,52 +63,80 @@ func initialize(ctxLogger *logrus.Entry) (*terraform.Executor, platform.Platform
+
+	switch bn {
+	case backend.Local:
+		bc, diags := local.NewConfig(&config.RootConfig.Backend.Config, config.EvalContext)
+		if diags.HasErrors() {
+			for _, diagnostic := range diags {
+				logger.Error(diagnostic.Error())
+			}
+
+			logger.Fatal("Errors found while loading backend configuration")
+		}
+
+		b, err := local.NewBackend(bc)
+		if err != nil {
+			logger.Fatalf("Error constructing backend: %v", err)
+		}
+
+		return b
```
>I suggest we focus to only select this unique information rather than re-doing the same thing for each backend.

What does this mean in practice? I don't understand your suggestion.

invidian replied:
I suggest we only select the right structure to valuate the config into, rather than re-doing error checking and creating the backend. So we have one line per backend type, not many.

johananl replied:
I understand. The problem is that I wanted to do all the "work" at creation time, i.e. unmarshaling, validation etc. It doesn't make sense to me to return an uninitialized, unvalidated struct and hope that the user calls all the relevant methods on the `Backend` interface.

AFAICT, to do what you're suggesting without exposing things like `LoadConfig()` and `Validate()` on the `Backend` interface we would have to do *everything* inside `NewBackend()`, which means we would have to return "regular" Go errors inside `hcl.Diagnostics` :confounded:

That's why I've chosen to separate the HCL unmarshaling from things like custom validations and rendering the template, and that's why we have a `Config` separately from the `Backend`, and *that's* why we have to explicitly create a `Config` and a `Backend` for each backend type.

johananl replied:
I'm wondering if we should stop using `hcl.Diagnostics` altogether. Personally I don't like this mechanism at all and would rather deal with "regular" errors myself.

invidian replied:
>I'm wondering if we should stop using hcl.Diagnostics altogether. Personally I don't like this mechanism at all and would rather deal with "regular" errors myself.

I'm for it, though I find mechanism for passing multiple errors useful, so we could have some replacement for it. Then we would just have to convert all returned diagnostics into our collection type.

Regarding the code, I would expect to have something like this as an implementation:
```go
func createBackend(logger *logrus.Entry, config *config.Config) backend.Backend {
  bn := config.RootConfig.Backend.Name

  var c interface{}

  switch bn {
  case backend.Local:
    c = &local.Config{}
  case backend.S3:
    c = &s3.Config{}
  }

  if d := gohcl.DecodeBody(config.RootConfig.Backend.Config, config.EvalContext, c); d.HasErrors() {
    logger.Fatalf("fail")
  }

  b, err := c.NewBackend()
  if err != nil {
    logger.Fatalf("fail")
  }

  return b
}
```

So the unique bit per backend is the struct holding the configuration. The rest of the code should be the same for all backends.

Also, I'd really like to not add new code, which use `Fatalf` for flow control - #630. We should be using logger for logging, not for the flow control.

johananl replied:
Also, what you're proposing kind of breaks encapsulation: now the caller needs to deal with `gohcl.DecodeBody()` which is an implementation detail of the backend AFAICT.

invidian replied:
>What concrete type does c.NewBackend() return?

It should return `Backend` interface, as in current PR (or specific type implementing backend interface).

>Also, what you're proposing kind of breaks encapsulation: now the caller needs to deal with gohcl.DecodeBody() which is an implementation detail of the backend AFAICT.

This is not the implementation detail of the backend. This is the UI part of the `lokoctl`. The backend should have no knowledge about HCL (maybe except struct tags).

johananl replied:
>It should return Backend interface, as in current PR (or specific type implementing backend interface).

I'm asking what **concrete** type you expect `NewBackend()` to return. I understand there is `backend.Backend` as the return type in the function's signature but you need to return a struct in the end.

>This is not the implementation detail of the backend. This is the UI part of the lokoctl. The backend should have no knowledge about HCL (maybe except struct tags).

My point is that it doesn't make sense to me to move `gohcl.DecodeBody()` to the UI. Does it make sense to you? Why?

invidian replied:
>My point is that it doesn't make sense to me to move gohcl.DecodeBody() to the UI. Does it make sense to you? Why?

In my opinion, we should isolate entire HCL code into one place, so it's only job is to convert HCL files into the Go struct, so we can operate on Go structs from there. This way, things are easier to test as test do not require HCL parser.

How about something like the following patch? Though still, in my opinion, having render method on config struct and return an error would even simplify it more.
```patch
From aeb00d3296deacf753ad5a6a33e64301ff51a47c Mon Sep 17 00:00:00 2001
From: Mateusz Gozdek <mateusz@kinvolk.io>
Date: Tue, 25 Aug 2020 18:31:15 +0200
Subject: [PATCH] patch

Signed-off-by: Mateusz Gozdek <mateusz@kinvolk.io>
---
 cli/cmd/cluster.go         | 44 +++++++++++---------------------------
 pkg/backend/backend.go     |  5 +++++
 pkg/backend/local/local.go | 14 ++++++------
 pkg/backend/s3/s3.go       | 12 ++++++-----
 4 files changed, 33 insertions(+), 42 deletions(-)

diff --git a/cli/cmd/cluster.go b/cli/cmd/cluster.go
index dd5d191c..9306905a 100644
--- a/cli/cmd/cluster.go
+++ b/cli/cmd/cluster.go
@@ -17,6 +17,7 @@ package cmd
 import (
        "fmt"

+       "github.com/hashicorp/hcl/v2/gohcl"
        "github.com/mitchellh/go-homedir"
        "github.com/sirupsen/logrus"
        "github.com/spf13/cobra"
@@ -85,44 +86,25 @@ func initialize(ctxLogger *logrus.Entry) (*terraform.Executor, platform.Platform
 func createBackend(logger *logrus.Entry, config *config.Config) backend.Backend {
        bn := config.RootConfig.Backend.Name

+       var backendConfig backend.Config
+
        switch bn {
        case backend.Local:
-               bc, diags := local.NewConfig(&config.RootConfig.Backend.Config, config.EvalContext)
-               if diags.HasErrors() {
-                       for _, diagnostic := range diags {
-                               logger.Error(diagnostic.Error())
-                       }
-
-                       logger.Fatal("Errors found while loading backend configuration")
-               }
-
-               b, err := local.NewBackend(bc)
-               if err != nil {
-                       logger.Fatalf("Error constructing backend: %v", err)
-               }
-
-               return b
+               backendConfig = &local.Config{}
        case backend.S3:
-               bc, diags := s3.NewConfig(&config.RootConfig.Backend.Config, config.EvalContext)
-               if diags.HasErrors() {
-                       for _, diagnostic := range diags {
-                               logger.Error(diagnostic.Error())
-                       }
-
-                       logger.Fatal("Errors found while loading backend configuration")
-               }
-
-               b, err := s3.NewBackend(bc)
-               if err != nil {
-                       logger.Fatalf("Error constructing backend: %v", err)
-               }
+               backendConfig = &s3.Config{}
+       }

-               return b
+       if d := gohcl.DecodeBody(config.RootConfig.Backend.Config, config.EvalContext, backendConfig); d.HasErrors() {
+               logger.Fatalf("bad backend")
        }

-       logger.Fatalf("Unknown backend %q", bn)
+       b, err := backendConfig.New()
+       if err != nil {
+               logger.Fatalf("bad backend")
+       }

-       return nil
+       return b
 }

 // initializeTerraform initialized Terraform directory using given backend and platform
diff --git a/pkg/backend/backend.go b/pkg/backend/backend.go
index 7136ae6b..f21e31da 100644
--- a/pkg/backend/backend.go
+++ b/pkg/backend/backend.go
@@ -26,3 +26,8 @@ type Backend interface {
        // String returns the Backend's configuration as rendered Terraform code.
        String() string
 }
+
+// Config describes functionality of backend configuration.
+type Config interface {
+       New() (Backend, error)
+}
diff --git a/pkg/backend/local/local.go b/pkg/backend/local/local.go
index 4af828a2..bb472926 100644
--- a/pkg/backend/local/local.go
+++ b/pkg/backend/local/local.go
@@ -19,7 +19,9 @@ import (

        "github.com/hashicorp/hcl/v2"
        "github.com/hashicorp/hcl/v2/gohcl"
+
        "github.com/kinvolk/lokomotive/internal/template"
+       "github.com/kinvolk/lokomotive/pkg/backend"
 )

 // Config represents the configuration of a local backend.
@@ -42,23 +44,23 @@ func NewConfig(b *hcl.Body, ctx *hcl.EvalContext) (*Config, hcl.Diagnostics) {
        return c, hcl.Diagnostics{}
 }

-// Backend implements the Backend interface for a local backend.
-type Backend struct {
+// backend implements the Backend interface for a local backend.
+type localBackend struct {
        config *Config
        // A string containing the rendered Terraform code of the backend.
        rendered string
 }

-func (b *Backend) String() string {
+func (b *localBackend) String() string {
        return b.rendered
 }

-// NewBackend constructs a Backend based on the provided config and returns a pointer to it.
-func NewBackend(c *Config) (*Backend, error) {
+// New constructs a Backend based on the provided config and returns a pointer to it.
+func (c *Config) New() (backend.Backend, error) {
        rendered, err := template.Render(backendConfigTmpl, c)
        if err != nil {
                return nil, fmt.Errorf("rendering backend: %v", err)
        }

-       return &Backend{config: c, rendered: rendered}, nil
+       return &localBackend{config: c, rendered: rendered}, nil
 }
diff --git a/pkg/backend/s3/s3.go b/pkg/backend/s3/s3.go
index 8ab4b2fd..7b450615 100644
--- a/pkg/backend/s3/s3.go
+++ b/pkg/backend/s3/s3.go
@@ -19,7 +19,9 @@ import (

        "github.com/hashicorp/hcl/v2"
        "github.com/hashicorp/hcl/v2/gohcl"
+
        "github.com/kinvolk/lokomotive/internal/template"
+       "github.com/kinvolk/lokomotive/pkg/backend"
 )

 // Config represents the configuration of an S3 backend.
@@ -67,18 +69,18 @@ func NewConfig(b *hcl.Body, ctx *hcl.EvalContext) (*Config, hcl.Diagnostics) {
 }

 // Backend implements the Backend interface for an S3 backend.
-type Backend struct {
+type s3Backend struct {
        config *Config
        // A string containing the rendered Terraform code of the backend.
        rendered string
 }

-func (b *Backend) String() string {
+func (b *s3Backend) String() string {
        return b.rendered
 }

-// NewBackend constructs a Backend based on the provided config and returns a pointer to it.
-func NewBackend(c *Config) (*Backend, error) {
+// New constructs a Backend based on the provided config and returns a pointer to it.
+func (c *Config) New() (backend.Backend, error) {
        if err := c.validate(); err != nil {
                return nil, fmt.Errorf("validating backend config: %w", err)
        }
@@ -88,5 +90,5 @@ func NewBackend(c *Config) (*Backend, error) {
                return nil, fmt.Errorf("rendering backend: %v", err)
        }

-       return &Backend{config: c, rendered: rendered}, nil
+       return &s3Backend{config: c, rendered: rendered}, nil
 }
--
2.28.0

```

johananl commented on pkg/backend/local/local.go:
```diff
@@ -15,40 +15,53 @@
-	backend.Register("local", NewLocalBackend())
-}
+// NewConfig creates a new Config and returns a pointer to it as well as any HCL diagnostics.
+func NewConfig(b *hcl.Body, ctx *hcl.EvalContext) (*Config, hcl.Diagnostics) {
```
That's what we had with `LoadConfig()`, however I don't think this is a good approach - see https://github.com/kinvolk/lokomotive/pull/828#discussion_r475761112.

johananl commented on pkg/backend/local/local.go:
```diff
@@ -15,40 +15,53 @@
-func (l *local) Validate() error {
-	return nil
+// NewBackend constructs a Backend based on the provided config and returns a pointer to it.
+func NewBackend(c *Config) (*Backend, error) {
```
I don't think this is extra work, you just do the error handling earlier. I don't see why we would want to allow creating a backend with a wrong config where you'll learn about the problem only when trying to render it.

johananl commented on pkg/backend/local/local.go:
```diff
@@ -15,40 +15,53 @@
+	if d := gohcl.DecodeBody(*b, ctx, c); len(d) != 0 {
+		diags = append(diags, d...)
+		return nil, diags
+	}
```
I've simplified this.

johananl committed:
Simplify backend handling

- Integrate the backend config into the root module template.
- Remove the Backend interface.

johananl committed:
platform/aks: Fix whitespace in template

The current whitespace trim markers cause the `locals` block to start
on the same line as the closing curly bracket of the previous block.

johananl commented:
I've made an experimental attempt to move the backend templates into the root module templates. I think that overall this simplifies things, however I still have to check how well this plays with refactoring the platform interface. This change introduces some new duplication which we should hopefully be able to get rid of while refactoring the platform interface.

I'm still checking whether this change makes sense. I've made the changes in a new commit which we can discard if we decide this approach isn't good.

invidian commented on pkg/backend/backend.go:
```diff
@@ -22,7 +31,48 @@ const (
-	String() string
+type Backend struct {
+	Type   string
+	Config interface{}
```
Why empty interface if we know all the types we deal with?

invidian commented on pkg/platform/aks/aks.go:
```diff
@@ -71,6 +73,9 @@ type config struct {
 	KubernetesVersion string
+
+	// TODO: Transient change - remove when refactoring platform interface.
+	Backend *backend.Backend
```
This is a noticeable change to the UI, since the block is being moved inside `cluster {}` block.

invidian commented on pkg/platform/aks/aks.go:
```diff
@@ -97,16 +102,35 @@ func init() { //nolint:gochecknoinits
-	if configBody == nil {
-		emptyConfig := hcl.EmptyBody()
-		configBody = &emptyConfig
+func (c *config) LoadConfig(cc *lkconfig.Config) hcl.Diagnostics {
```
As we need to get rid of `LoadConfig()` completely, I'm not sure why we add more code into it now. Also accepting entire configuration here breaks the encapsulation :confused:

invidian reviewed (changes requested):
>I've made an experimental attempt to move the backend templates into the root module templates. I think that overall this simplifies things, however I still have to check how well this plays with >refactoring the platform interface. This change introduces some new duplication which we should hopefully be able to get rid of while refactoring the platform interface.

>I'm still checking whether this change makes sense. I've made the changes in a new commit which we can discard if we decide this approach isn't good.

Hmm, I don't think those changes go in the right direction to be honest. They bring UI changes, empty interfaces, break the encapsulation of Platform code, add duplication and spread HCL loading even more by expanding function which we want to get rid of :confused: 

@johananl what is wrong with changes I suggested in https://github.com/kinvolk/lokomotive/pull/828#discussion_r476603787 ?
